opendevin/runtime/plugins/jupyter/__init__.py

```diff
@@ -29,33 +30,39 @@ class JupyterRequirement(PluginRequirement):
 class JupyterPlugin(Plugin):
     name: str = 'jupyter'
 
-    def initialize(self, kernel_id: str = 'opendevin-default'):
+    async def initialize(self, username: str, kernel_id: str = 'opendevin-default'):
         self.kernel_gateway_port = find_available_tcp_port()
         self.kernel_id = kernel_id
         self.gateway_process = subprocess.Popen(
-            [
-                '/opendevin/miniforge3/bin/mamba',
-                'run',
-                '-n',
-                'base',
-                'poetry',
-                'run',
-                'jupyter',
-                'kernelgateway',
-                '--KernelGatewayApp.ip=0.0.0.0',
-                f'--KernelGatewayApp.port={self.kernel_gateway_port}',
-            ],
+            (
+                f"su - {username} -s /bin/bash << 'EOF'\n"
+                'cd /opendevin/code\n'
+                'export POETRY_VIRTUALENVS_PATH=/opendevin/poetry;\n'
+                '/opendevin/miniforge3/bin/mamba run -n base '
+                'poetry run jupyter kernelgateway '
+                '--KernelGatewayApp.ip=0.0.0.0 '
+                f'--KernelGatewayApp.port={self.kernel_gateway_port}\n'
+                'EOF'
+            ),
             stderr=subprocess.STDOUT,
+            shell=True,
         )
         # read stdout until the kernel gateway is ready
+        output = ''
         while True and self.gateway_process.stdout is not None:
             line = self.gateway_process.stdout.readline().decode('utf-8')
+            output += line
             if 'at' in line:
                 break
             time.sleep(1)
-            print('Waiting for jupyter kernel gateway to start...')
+            logger.debug('Waiting for jupyter kernel gateway to start...')
 
-    async def run(self, action: Action) -> Observation:
+        logger.info(
+            f'Jupyter kernel gateway started at port {self.kernel_gateway_port}. Output: {output}'
+        )
+        await self.run(IPythonRunCellAction(code='import os; os.chdir("/workspace")'))
```

Comment:
Should this be `/workspace`? Maybe it needs to be `config.workspace_path_in_sandbox`?

Also related to https://github.com/OpenDevin/OpenDevin/issues/3173, maybe we can update this when bash changes the directory, although this could obviously be in a future PR?

Reply:
Yep - we have similar stuff here, though it only updates the `JUPYTER_PWD` env var, which only sync to pwd of all agentskills library tools

https://github.com/OpenDevin/OpenDevin/pull/3157/files/927a77dcbe467e027b8e615580aa5f94b596f68d#diff-4c49958e502b9e841881ad39debec1f606220fe193624e53e280b3beee0830fbR234-R245

Reply:
Sounds good @xingyaoww , but I still think that within this PR this probably should be fixed to consider `workspace_mount_path_in_sandbox`:

https://github.com/OpenDevin/OpenDevin/blob/2953438c98555d7dd632d6b057dbceae11708e8b/opendevin/core/config.py#L207

Reply:
done! This Plugin is actually running in the client side, so i chdir to `self.pwd`, which is initially set to `workspace_mount_path_in_sandbox`:
https://github.com/OpenDevin/OpenDevin/pull/3157/commits/3067553f3ce5f88939050e9df824bc2f8749279c